ui/nuxeo-document-layout.js

```diff
@@ -0,0 +1,303 @@
+     */
+    validate() {
+      return this.$.layout.validate();
+    }
+
+    /**
+     * Applies focus on the first layout child with the `autofocus` attribute.
+     *
+     * Note: This method won't do anything if there's already a child element focused.
+     */
+    applyAutoFocus() {
+      const focusableElement = this._getFocusableElement(this.element);
+      if (focusableElement) {
+        focusableElement.focus();
+      }
+    }
+
+    /**
+     * Displays any errors present in a validation report, and invalidates layout widgets bound to document
+     * fields whose constraints were violated.
+     * @param {object} validationReport The validation report to be displayed.
+     */
+    reportValidation(validationReport) {
+      this._resetValidationErrors();
+      validationReport.violations.reverse().forEach((violation) => {
+        this.invalid = true;
+        if (violation.path) {
+          violation.path.forEach((p) => {
+            const widgets = this._getBoundElements(`document.properties.${p.field_name}`);
+            if (widgets) {
+              const msg = this.i18n(violation.messageKey, violation.invalid_value, p.field_name);
+              if (msg === violation.messageKey && violation.constraint && violation.constraint.name) {
+                this._addValidationError(
+                  this.i18n(
+                    `label.schema.constraint.violation.${violation.constraint.name}`,
+                    violation.invalid_value,
+                    p.field_name,
+                    ...Object.values(violation.constraint.parameters),
+                  ),
+                );
+              } else {
+                this._addValidationError(msg);
+              }
+              Object.values(widgets).forEach((widget) => {
+                // we can at least flag the widget `invalid`
+                widget.invalid = true;
+              });
+            } else {
+              this._addValidationError(this.i18n(violation.messageKey, violation.invalid_value, p.field_name));
+            }
+          });
+        } else {
+          this._addValidationError(this.i18n(violation.messageKey));
+        }
+      });
+    }
+
+    _buildHrefFn(tmpl) {
+      // eslint-disable-next-line no-new-func
+      return new Function(['document', 'layout', 'hrefBase'], `return \`${tmpl}\`.toLowerCase();`);
+    }
+
+    _loadLayout(document, layout, hrefFunction) {
+      this._resetValidationErrors();
+      if (document) {
+        if (!this.previousDocument || document.uid !== this.previousDocument.uid) {
+          this._set_href(null); // force layout restamp
+        }
+        if (!this.previousDocument || document.type === this.previousDocument.type) {
+          this._set_model({ document });
+        }
+        const path = hrefFunction(document, layout, this.hrefBase);
+        this._set_href(this.resolveUrl(path, (this.__dataHost && this.__dataHost.importPath) || this.importPath));
```

Comment:
the second parameter here is the base path, this is where the `hrefBase` should be used. By default it's the `importPath` of the data host.

Reply:
I was assuming the href base would be in relation to the importPath, but it doesn't have to be that way. Perhaps it can be used directly on the resolveUrl.